### DISCUSSION TITLE
Selenium module: VNC recorder doesn't stop recording when test ends

### DISCUSSION THREAD

**Discussion by deejgregor** [testcontainers/testcontainers-java#6229] [Ideas]
Today, if VNC recording is enabled in `BrowserWebDriverContainer`, we create the `vncRecordingContainer` and [start](https://github.com/testcontainers/testcontainers-java/blob/main/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L303) it. Once the test is done, if we retain the recording (in [afterTest](https://github.com/testcontainers/testcontainers-java/blob/main/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L346) and [retainRecordingIfNeeded](https://github.com/testcontainers/testcontainers-java/blob/main/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L373)), the VNC recording isn't stopped before transcoding is started. Recording is normally only stopped in [stop](https://github.com/testcontainers/testcontainers-java/blob/main/modules/selenium/src/main/java/org/testcontainers/containers/BrowserWebDriverContainer.java#L363), which doesn't happen until after we retain the recording.

In many cases this isn't too much of an issue as long as the transcoding with `ffmpeg` can process the recording faster than newer data is received. In this case, the screen will still be continue to be recorded while transcoding is happening, but eventually transcoding catches up, hits EOF on the input file and `ffmpeg` exits, at which point the file is copied and later the containers are stopped.

I ran into two cases where this doesn't happen:
* On an arm64-based system running the `vnc-recording` container, which is available for amd64 only, `ffmpeg` runs super slow and never has any hope of catching up. I'm working on a PR to make versions for other architectures. See: https://github.com/testcontainers/vnc-recorder/pull/5
* In OpenNMS, we have a fairly massive screen size (2048x1400), and at this size, the incoming data rate is higher than `ffmpeg`'s processing rate, even when the container architecture matches the host architecture. I can reproduce this with example `selenium-container` if I increase the screen size.

When the transcoder can't keep up, the `retainRecordingIfNeeded` method never returns and the test never completes.

I was able to workaround this by sending a SIGSTOP to the VNC recorder process inside its Docker container. Note that we can't just stop the container because we need it around and running since that's where the transcoder runs--at least, that's where it runs today. You can see [my workaround for OpenNMS](https://github.com/OpenNMS/opennms/blob/develop/smoke-test/src/test/java/org/opennms/smoketest/OpenNMSSeleniumIT.java#L214).

A backward-compatible fix with existing `vnc-recording` containers would be to put a similar SIGSTOP solution in place, probably in `afterTest`, I think (regardless of whether or not we want to retain the recording, I think it makes sense to stop recording in any event).

A cleaner approach would be to add some more specific support to the container to stop recording (maybe a script that tells the python recording process to stop recording but stick around so the container doesn't stop). But this would require both vnc-recorder and testcontainers-java (and other language versions?) to be updated. My suggestion would be to first get the fix in place that works with existing containers. Work on a cleaner approach could start after that, if desired.

I'm happy to submit a PR with a cleaned up version of the SIGSTOP approach mentioned above along with a test case. I'm also very happy to take suggestions for other approaches and anything else you would like to be done with this. Note: it might make sense to add a (configurable?) delay before stopping recording.

**Comment by deejgregor** [comment]
Here are a few cases that I've seen that seem to be related to this:
* https://github.com/testcontainers/testcontainers-java/issues/5268 / https://github.com/testcontainers/testcontainers-java/pull/5269
* https://github.com/testcontainers/testcontainers-java/issues/5024

**Comment by kiview** [comment]
Thanks a lot for your detailed explanation of the issue and the suggestions @deejgregor. I agree completely with your proposal and the following approach seems sensible:

1. Add a `SIGSTOP` like implementation to `afterTest()`, this is a very non-invasive change and will likely also fix the mentioned issues.
2. Optionally: Implement cleaner support for this function to `vnc-recorder`. But I'd say, if 1) already gives us good results, maybe no need for this additional work.

We'd be very happy if you can provide a PR for 1 🙂

**Reply by deejgregor** [reply]
PR #6307 opened.

**Comment by AB-xdev** [comment]
So after searching for a few weeks why our tests in our CI got randomly stuck and run into timeouts, after a lot of debugging I found out that this issue seems to cause the problems (VNC recorder was consuming all resources until it died from memory starvation).

As a consequence (we also found other problems) we re-implemented the video recording using [Seleniums native video recorder](https://github.com/SeleniumHQ/docker-selenium/blob/trunk/README.md#video-recording).
The [corresponding library is now available as open source](https://github.com/xdev-software/testcontainers-selenium), feel free to use it.
